### PR TITLE
[WIP] Experiment with ReSwift

### PR DESCRIPTION
**Pull request by ashfurrow**
This lines up with the work I've been doing on prediction, as well as the ReSwift thing I'm doing at The Swift Alps next week.

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>

    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="true">No CHANGELOG changes made</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>

    <tr>
      <td>:warning:</td>
      <td data-sticky="true">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Needs testing on a Phone if change is non-trivial</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="true">Detected some Swift building time outliers</td>
    </tr>
  </tbody>
</table>

Time | Class | Function |
| --- | ----- | ----- |257.7ms | [AuctionViewController.swift](/artsy/eigen/tree/ashfurrow-reswift/Artsy/View_Controllers/Auction/AuctionViewController.swift#L199) | @objc dynamic func showRefineTappedAnimated(animated: Bool)
209.2ms | [SwiftyJSON.swift](/artsy/eigen/tree/ashfurrow-reswift/Pods/SwiftyJSON/Source/SwiftyJSON.swift#L445) | get {}
180.3ms | [WebSocket.swift](/artsy/eigen/tree/ashfurrow-reswift/Pods/Starscream/Source/WebSocket.swift#L528) | private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)
95.3ms | [LiveAuctionLotViewController.swift](/artsy/eigen/tree/ashfurrow-reswift/Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift#L47) | @objc override func viewDidLoad()
85.5ms | [Observable.swift](/artsy/eigen/tree/ashfurrow-reswift/Pods/Interstellar/Sources/Observable.swift#L71) | private final func nextTokenHash() -> Int
79.6ms | [AuctionLotMetadataStackScrollView.swift](/artsy/eigen/tree/ashfurrow-reswift/Artsy/View_Controllers/Live_Auctions/Views/AuctionLotMetadataStackScrollView.swift#L18) | required init(viewModel: LiveAuctionLotViewModelType, salesPerson: LiveAuctionsSalesPersonType, sideMargin: String)
75.9ms | [LiveAuctionsAdminViewController.swift](/artsy/eigen/tree/ashfurrow-reswift/Artsy/View_Controllers/Live_Auctions/LiveAuctionsAdminViewController.swift#L22) | @objc override func viewDidLoad()
68.6ms | [UIViewController+BlurredStatusView.swift](/artsy/eigen/tree/ashfurrow-reswift/Artsy/View_Controllers/Core/UIViewController+BlurredStatusView.swift#L25) | func ar_presentBlurredOverlayWithTitle(title: String, subtitle: String, buttonState: BlurredStatusOverlayViewCloseButtonState = default)
66.6ms | [UIViewController+BlurredStatusView.swift](/artsy/eigen/tree/ashfurrow-reswift/Artsy/View_Controllers/Core/UIViewController+BlurredStatusView.swift#L31) | (closure)


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: Artsy/View_Controllers/Live_Auctions/IsInitialStateLoaded.swift [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import ReSwift
+
+func isInitialStateLoaded(state: Bool?, action: Action) -> Bool {
```
Comment (acjay) [inline Artsy/View_Controllers/Live_Auctions/IsInitialStateLoaded.swift]: Ha, ReSwift is pretty literally the same as Redux, eh?

**Comment by acjay** [comment]
Occurs to me, given that we're using JS in the app elsewhere, I wonder if it's at all feasible to factor out the state management from Refraction into its own module, and then completely share the code between the two projects?

**Comment by ashfurrow** [comment]
@acjay Did you mean Refraction or Prediction? From Prediction's side, I'm thinking the main benefits would be from the state management. I'd like to discuss the kinds of benefits we could get, especially since it would help modularize the Redux store state away from being tied to the Prediction UI specifically. I'm going to of some exploration in the mean time.

**Comment by acjay** [comment]
LOL, oops, yeah, Prediction.

Actually, I think Refraction is a good candidate for replacement by a new in-room display mode in Prediction. Although I can't see that being a top priority any time soon.

**Comment by ashfurrow** [comment]
Cool, let's chat about this on Wednesday 👍 Eloy has offered his guidance on interoperating with JS code, really excited to evaluate what this could potentially help us with.

**Comment by ashfurrow** [comment]
Talked with Alan about this and we agree on principle that it's possible, I'm looking into specifics of what could be made redundant by code sharing. 

Our `LiveAuctionsSalesPerson` currently maintains a lot of state and distributes updates via Interstellar `Observable`s, a lot of this could be replaced by Redux-style selectors and some kind of state update mechanism. I believe that the existing LAI user interface is sufficiently component-ish-based that moving to a different subscription model wouldn't be too arduous (that's why I opened this PR in the first place). The `LiveAuctionStateManager`, `LiveAuctionStateReconciler`, and `LiveAuctionSocketCommunicator` classes would be completely replaced, though the `LiveAuctionStaticDataFetcher` which handles the Metaphysics call would probably remain. 

I'm not 100% clear on next steps, I'll schedule a meeting to discuss 👍

**Comment by orta** [comment]
This is going to be considerably broken by the Xcode8 merge

**Comment by ashfurrow** [comment]
Aye, given the interest in sharing code with prediction, I'm going to close this PR. Even if we do end up using ReSwift, it'll be easier to just re-do this PR.

**Comment by manojmahapatra** [comment]
@ashfurrow @orta Do you know what's current status of this PR? Basically we are considering to use `reswift` in one our projects and I wanted to know how to leverage this to its full potential. Are there any known/unknown things to consider before kickstarting and/or using the redux style approach? e.g. how the custom router works while supporting for universal linking, persisting and restoring of navigation stacks etc.

Appreciate for any help.

**Comment by ashfurrow** [comment]
@manojmahapatra great question. We dropped this because we didn't have time to make it work. We investigated bridging with our web's Redux store but again, time constraints stopped us from doing it. There are no issues that I'm aware of but I'm afraid I haven't used ReSwift in a large production-scale app. Let me know if I can clarify anything!

**Comment by manojmahapatra** [comment]
@ashfurrow basically I wanted to know/see if ReSwift been used in any of the large production-scale app. What are the challenges if anyone has faced, design decisions etc that been taken during the development phase since we're considering to use it in a substantially large app. So i think its always better to do all the research before start implementing w/ anything.

**Comment by ashfurrow** [comment]
> its always better to do all the research before start implementing w/ anything

Very wise, basically the opposite of my approach :joy: 

I've not heard of ReSwift being used in large open source apps, or large apps in general, but we use the Redux flow for our React web apps and it's quite nice 👍

**Pull request closed without merging**